### PR TITLE
fix: stub custom node kinds during schemaless ingest [BED-7924] 

### PR DESCRIPTION
**Pull request by kpom-specter**
## Description
stub custom node kinds during schemaless ingest

## Motivation and Context

BED-7924

*Why is this change required? What problem does it solve?*

## How Has This Been Tested?

*Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc.*

## Types of changes

- Chore (a change that does not modify the application functionality)
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to change)
- Database Migrations

## Checklist:

- [ ] I have met the contributing prerequisites
  - Assigned myself to this PR
  - Added the appropriate labels
  - Associated an issue: https://github.com/SpecterOps/BloodHound/issues/672
  - Read the Contributing guide: https://github.com/SpecterOps/BloodHound/wiki/Contributing
- [ ] I have ensured that related documentation is up-to-date
  - Open API docs
  - Code comments (GoDocs / JSDocs)
- [ ] I have followed proper test practices
  - Added/updated tests to cover my changes
  - All new and existing tests passed

## Summary by CodeRabbit

* **New Features**
  * Dynamic registration of custom node kinds during ingest with automatic stub creation and in-memory refresh.
  * Added an ingest-time API to ensure stubbed custom node kinds are present.

* **Bug Fixes**
  * Relationship ingestion now only runs when relationship data exists.

* **Tests**
  * Added integration and unit tests covering registration, idempotency, error handling, and edge cases.

* **Chores**
  * Database migration to add the ingest-time stubbed-kind helper.

[Review Change Stack](https://app.coderabbit.ai/change-stack/SpecterOps/BloodHound/pull/2770)

### CODE REVIEW

File: cmd/api/src/database/customnode_integration_test.go [hunk 1]
```diff
@@ -523,3 +591,21 @@ func TestDeleteCustomNodeKind(t *testing.T) {
 		})
 	}
 }
+
+func countKindRows(t *testing.T, testSuite IntegrationTestSuite, kindName string) int64 {
+	t.Helper()
+
+	var count int64
+	result := testSuite.DB.WithContext(testSuite.Context).Raw("SELECT COUNT(*) FROM kind WHERE name = ?", kindName).Scan(&count)
```
Comment (mistahj67) [inline cmd/api/src/database/customnode_integration_test.go]: Can this and the below leverage the model.TableName() methods respectively? This will ensure if ever the table changes these don't break

File: cmd/api/src/services/graphify/ingest.go [hunk 2]
```diff
@@ -243,17 +253,55 @@ func IngestBasicData(batch *IngestContext, converted ConvertedData) error {
...
+	}
+
+	if batch.seenKinds == nil {
+		batch.seenKinds = map[string]struct{}{}
+	}
+
+	for _, node := range nodes {
+		for _, kind := range MergeNodeKinds(sourceKind, node.Labels...) {
+			kindName := kind.String()
+			if strings.HasPrefix(kindName, "Tag_") {
```
Comment (mistahj67) [inline cmd/api/src/services/graphify/ingest.go]: I believe there is a constant for `Tag_` but there are additional internal node kinds that are ignored, the logic currently lives in `kinds` handler layer but can be moved to the model layer to leverage both here as well as in https://github.com/SpecterOps/BloodHound/pull/2766

File: cmd/api/src/database/migration/migrations/20260513120000_stub_custom_node_kind_for_ingest.sql [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+BEGIN
+  -- Mirrors upsert_kind's locking behavior without burning sequence values on conflicts.
+  PERFORM pg_advisory_xact_lock(hashtext(node_kind_name));
+
+  SELECT * INTO kind_row FROM kind WHERE name = node_kind_name;
+  IF kind_row.id IS NOT NULL THEN
+    RETURN;
+  END IF;
+
+  INSERT INTO kind (name) VALUES (node_kind_name) RETURNING * INTO kind_row;
```
Comment (mistahj67) [inline cmd/api/src/database/migration/migrations/20260513120000_stub_custom_node_kind_for_ingest.sql]: Would it make sense to call the `upsert_kind` function here instead? That way any changes to that function propagate to this one in any future updates?

The double select is not ideal but might be worth the tradeoff of maintaining "2" routes to insert kinds

File: cmd/api/src/services/graphify/ingest.go (hunk 2 above)
Comment (kpom-specter) [reply cmd/api/src/services/graphify/ingest.go]: This makes a ton of sense; 2766 will want to update once this merges

File: cmd/api/src/database/migration/migrations/20260513120000_v9_stub_custom_node_kind_for_ingest.sql
Comment (mistahj67) [inline cmd/api/src/database/migration/migrations/20260513120000_v9_stub_custom_node_kind_for_ingest.sql:1]: It has been brought to my attention that the major version number in these file names is required: 
`20260513120000_v9_stub_custom_node_kind_for_ingest.sql`

**Review by mistahj67 — CHANGES_REQUESTED** [review]

File: cmd/api/src/database/customnode.go [hunk 4]
```diff
@@ -24,17 +24,39 @@ import (
...
 	CreateCustomNodeKinds(ctx context.Context, customNodeKind model.CustomNodeKinds) (model.CustomNodeKinds, error)
+	EnsureStubbedCustomNodeKindForIngest(ctx context.Context, name string) error
 	GetCustomNodeKinds(ctx context.Context, filters model.Filters) ([]model.CustomNodeKind, error)
 	GetCustomNodeKind(ctx context.Context, kindName string) (model.CustomNodeKind, error)
 	UpdateCustomNodeKind(ctx context.Context, customNodeKind model.CustomNodeKind) (model.CustomNodeKind, error)
 	DeleteCustomNodeKind(ctx context.Context, kindName string) error
 }
 
+func (s *BloodhoundDB) EnsureStubbedCustomNodeKindForIngest(ctx context.Context, name string) error {
+	if name == "" {
```
Comment (mistahj67) [inline cmd/api/src/database/customnode.go]: What do you think of tossing the kind through the `model.IsExtendedNodeKind(graph.StringKind(name))` and if it passes, you kick out the same error? That way there's another layer of protection around those kinds

File: cmd/api/src/model/kind.go [hunk 5]
```diff
@@ -16,7 +16,19 @@
...
+
+	"github.com/specterops/bloodhound/packages/go/graphschema/common"
+	"github.com/specterops/dawgs/graph"
+)
+
+// IsExtendedNodeKind reports whether a kind is an internal or framework kind
+// that should not be treated as a user-ingested node kind.
+func IsExtendedNodeKind(kind graph.Kind) bool {
+	return strings.HasPrefix(kind.String(), AssetGroupTagKindPrefix) ||
+		kind.Is(common.MigrationData, graph.StringKind("Meta"), graph.StringKind("MetaDetail"))
```
Comment (mistahj67) [inline cmd/api/src/model/kind.go]: Suggestion: I believe these have been formally added to the graphschema pkg and you can grab them there now.

**Review by mistahj67 — APPROVED** [review]
Tested locally, looks to address the concern. Had a couple small possible improvements but not blocking! Thanks for diving in and addressing

File: cmd/api/src/database/customnode.go (hunk 4 above)
Comment (kpom-specter) [reply cmd/api/src/database/customnode.go]: Yeah, that makes a lot of sense. Good catch!
